utils/bitfield/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+        match other {
+            BitField::Encoded { bv, set, unset } => {
+                *self.as_mut_flushed()? &= decode_and_apply_cache(bv, set, unset)?
+            }
+            BitField::Decoded(bv) => *self.as_mut_flushed()? &= bv.iter().copied(),
+        }
+        Ok(())
+    }
+
+    /// Subtract other bitfield from self (equivalent of `a & !b`)
+    pub fn subtract(mut self, other: &Self) -> Result<Self> {
+        self.subtract_assign(other)?;
+        Ok(self)
+    }
+
+    /// Subtract other bitfield from self (equivalent of `a & !b`)
+    pub fn subtract_assign(&mut self, other: &Self) -> Result<()> {
+        match other {
+            BitField::Encoded { bv, set, unset } => {
+                *self.as_mut_flushed()? &= !decode_and_apply_cache(bv, set, unset)?
+            }
+            BitField::Decoded(bv) => *self.as_mut_flushed()? &= bv.iter().copied().map(|b| !b),
+        }
+        Ok(())
+    }
+
+    /// Creates a bitfield which is a union of a vector of bitfields.
+    pub fn union(bit_fields: &[Self]) -> Result<Self> {
+        let mut ret = Self::default();
+        for bf in bit_fields.iter() {
+            ret.merge_assign(bf)?;
+        }
+        Ok(ret)
+    }
+
+    /// Returns true if BitFields have any overlapping bits.
+    pub fn contains_any(&mut self, other: &mut BitField) -> Result<bool> {
+        for (&a, &b) in self
+            .as_mut_flushed()?
+            .iter()
+            .zip(other.as_mut_flushed()?.iter())
+        {
+            if a && b {
+                return Ok(true);
+            }
+        }
+        Ok(false)
+    }
+
+    /// Returns true if the self `BitField` has all the bits set in the other `BitField`.
+    pub fn contains_all(&mut self, other: &mut BitField) -> Result<bool> {
+        let a_bf = self.as_mut_flushed()?;
+        let b_bf = other.as_mut_flushed()?;
+
+        // Checking lengths should be sufficient in most cases, but does not take into account
+        // decoded bitfields with extra 0 bits. This makes sure there are no extra bits in the
+        // extension.
+        if b_bf.len() > a_bf.len() && b_bf[a_bf.len()..].count_ones() > 0 {
+            return Ok(false);
+        }
+
+        for (a, b) in a_bf.iter().zip(b_bf.iter()) {
+            if *b && !a {
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+}
+
+fn bit_or<I>(a: &mut BitVec, mut b: I)
+where
+    I: Iterator<Item = bool>,
+{
+    for mut a_i in a.iter_mut() {
+        match b.next() {
+            Some(true) => *a_i = true,
+            Some(false) => (),
+            None => return,
+        }
+    }
+
+    a.extend(b);
+}
+
+fn decode_and_apply_cache(
+    bit_vec: &BitVec,
+    set: &FnvHashSet<u64>,
+    unset: &FnvHashSet<u64>,
+) -> Result<BitVec> {
+    let mut decoded = rleplus::decode(bit_vec)?;
+
+    // Resize before setting any values
+    if let Some(&max) = set.iter().max() {
+        decoded.resize(max as usize + 1, false);
```

Comment:
I think we should only resize the bitvec here if `max + 1` actually exceeds the length of the existing bitvec? Otherwise it gets truncated